SQLAlchemy Models for Blue Core Resources with Triples Embedding

jermnelson opened the pull request:
## Why was this change made?
Fixes #1, Fixes #10, Fixes #19

Based on previous draft PRs #23 and #26 and comments in slack from @edsu. 

Things to note:
- These SQLAlchemy models use polymorphic parent/child classes with a parent `ResourceBase` class for both JSON-LD and JSON resources managed by Blue Core. 
- Two child classes of `ResourceBase`, `Work` and `Instance`, are for BIBFRAME Works and Instances that actively managed by Blue Core
- The third child class of `ResourceBase`, `OtherResource`, stores RDF resources that are in the RDF record but aren't BIBFRAME Works or Instances. This class can also store pure JSON profiles to support Marva and JSON-LD templates to support Sinopia. `OtherResource` are associated with either `Work` or `Instance` objects through the `BibframeOtherResources` class.
- The `Version` class stores the BIBFRAME Work and Instances JSON-LD with new `Version` being added when a `Work` or `Instance` is updated.
- The `BibframeClass` and `ResourceBibframeClass` classes are used to associate one or more BIBFRAME Work or Instance classes with either the `Work` or `Instance`
- The `TripleVectorIndex` stores an embedding vector for each triple in a `Version`

edsu commented:
I think this will help me at least when looking at the schema. You did it already but I wonder if it might be worth sketching out what we expect to be of type OtherResource, and if it might make sense to split those out as separate tables now. I know that this is kind of against the grain of graph database where everything is reduced down to triples, but we are using a relational database so it might be good to be explicit about the types of entities we are going to be modeling?

I noticed when following along with the README to install that I hit a couple snags:

```
To use the fastapi command, please install "fastapi[standard]":

        pip install "fastapi[standard]"

Traceback (most recent call last):
  File "/Users/edsu/Projects/blue-core-lod/bluecore_api/.venv/bin/fastapi", line 8, in <module>
    sys.exit(main())
             ~~~~^^
  File "/Users/edsu/Projects/blue-core-lod/bluecore_api/.venv/lib/python3.13/site-packages/fastapi/cli.py", line 12, in main
    raise RuntimeError(message)  # noqa: B904
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: To use the fastapi command, please install "fastapi[standard]":

        pip install "fastapi[standard]"
```

When I updated the pyproject.toml to include `fastapi[standard]` instead of `fastapi` I then ran into:

```
➜  bluecore_api git:(postgres-bf-models) ✗ uv run fastapi dev src/bluecore/app/main.py --port 3000
warning: `uv run` is experimental and may change without warning

   FastAPI   Starting development server 🚀

             Searching for package file structure from directories with __init__.py files

             Path does not exist src/bluecore/app/main.py
```

I guess main.py moved somewhere else?

jermnelson commented:
> Thanks! I think this will help me at least when looking at the schema and working with the database abstractions.
> 
> You did it a bit already but I wonder if it might be worth sketching out what we expect to be of type `OtherResource`, and if it might make sense to split those out as separate tables now. I know that this is kind of against the grain of graph database where everything is reduced down to triples, but we are using a relational database so it might be good to be explicit about the types of entities we are going to be modeling? If this incurs too much work at this point I'm willing to punt on it, but I just thought it might be worth talking about.
> 
> I noticed when following along with the README to install that I hit a couple snags:
> 
> ```
> To use the fastapi command, please install "fastapi[standard]":
> 
>         pip install "fastapi[standard]"
> 
> Traceback (most recent call last):
>   File "/Users/edsu/Projects/blue-core-lod/bluecore_api/.venv/bin/fastapi", line 8, in <module>
>     sys.exit(main())
>              ~~~~^^
>   File "/Users/edsu/Projects/blue-core-lod/bluecore_api/.venv/lib/python3.13/site-packages/fastapi/cli.py", line 12, in main
>     raise RuntimeError(message)  # noqa: B904
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> RuntimeError: To use the fastapi command, please install "fastapi[standard]":
> 
>         pip install "fastapi[standard]"
> ```
> 
> When I updated the pyproject.toml to include `fastapi[standard]` instead of `fastapi` I then ran into:
> 
> ```
> ➜  bluecore_api git:(postgres-bf-models) ✗ uv run fastapi dev src/bluecore/app/main.py --port 3000
> warning: `uv run` is experimental and may change without warning
> 
>    FastAPI   Starting development server 🚀
> 
>              Searching for package file structure from directories with __init__.py files
> 
>              Path does not exist src/bluecore/app/main.py
> ```
> 
> I guess main.py moved somewhere else?

I updated the README to reflect that the API is still in development. The README now includes directions to bring-up the Docker Postgres Database and run the Alembic migrations. The README also includes directions for the running the model unit tests and linting with ruff.

edsu commented:
@jermnelson and I chatted about potential risks to using SQLAlchemy's [polymorphic class inheritance](https://docs.sqlalchemy.org/en/20/orm/inheritance.html#building-deeper-hierarchies-with-polymorphic-abstract) in terms of migrating the schema going forward. Usually inheritance can pose problems when refactoring, especially when it isn't strictly necessary, and it also can make the schema more difficult to understand. Jeremy suggested that it would be helpful to try a Alembic migration to the schema involving one of the involved models and see how well it works. Based on that I think we will know whether we want to move forward this way.

jermnelson commented:
Following up on @edsu [comment](https://github.com/blue-core-lod/bluecore_api/pull/27#issuecomment-2620289317) comment,  the [test-migration](https://github.com/blue-core-lod/bluecore_api/tree/test-migration) branch is an example migration.  To test actually data migration, three LOC records were loaded before the following changes.

In [models.py](https://github.com/blue-core-lod/bluecore_api/blob/test-migration/src/bluecore/models.py) added a `title` field to the Work with nullable set to False:

```python
class Work(ResourceBase):
    __tablename__ = "works"
   ...
   title: Mapped[str] = mapped_column(String, nullable=False)
```

For the Instance model added a `generation` field:
```python
class Instance(ResourceBase):
    __tablename__ = "instances"
   ....
    generation: Mapped[str] = mapped_column(String, nullable=True)
```

After running the `uv run alembic revision` to auto-generate the basic scaffolding for a migration, running `uv run alembic upgrade head` failed because the three test Work rows required a title value. To fix these errors required adjusted and adding the following to the [upgrade](https://github.com/blue-core-lod/bluecore_api/blob/07e54374e4dae05a4eb9ebe0b8adb70d4f40feb5/bluecore_store_migrations/versions/2486cd208dd9_adds_properties_to_work_and_instance.py#L23) function.

```python
def upgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    op.add_column('instances', sa.Column('generation', sa.String(), nullable=True))
    op.add_column('works', sa.Column('title', sa.String(), nullable=True))
    # ### end Alembic commands ###
    connection = op.get_bind()
    result = connection.execute(sa.text("SELECT id, data FROM resource_base WHERE type = 'works';"))
    for id, data in result:
        graph = rdflib.Graph()
        graph.parse(data=data, format="json-ld")
        title = next(graph.objects(predicate=rdflib.URIRef("http://id.loc.gov/ontologies/bibframe/mainTitle")))
        if title is None:
            title = "Title Not Found"
        print(f"Updating Work {id} with title {title}")
        connection.execute(
            sa.text("UPDATE works SET title = :title WHERE id = :id;"),
            { "title": title,
              "id": id }
        )
    # Restore nullable constraint in Work model
    op.alter_column('works', 'title', existing_type=sa.String(), nullable=False)
```

1.  Changed the `title` column to `nullable=True`
2. Queried for the existing works to extract the id and rdf data
3. Find the BIBFRAME mainTitle from the RDF and provide a default value if not present
4. Updated the works table with the new title 
5. Altered the works table to set back to `nullable=False`

Running the `uv run alembic upgrade head` now succeeds. While it would be nice if Alembic would automagically create the data migration code, that would have been difficult because of the RDF extraction and parsing. The resulting Work and Instance objects contained the new properties with the `Work.title` field populated with JSON-LD BF.mainTitle value. The only explicit polymorphic class was in the first SQL query to extract the `id` and `data` fields, the remainder of database interactions were directly on the `works` table.

edsu commented:
I think we need to add psycopg2 to the dependencies?

Also maybe we could remove `docker-compose.yml` if the Mongo config is no longer needed. I was super confused at first when the Postgres connection failed when I ran `uv run alembic upgrade head` and got a connection error. I thought maybe Mongo was running, but in reality the Dockerized PostgreSQL from the `compose.yml` was running it just couldn't bind to the port as I had PostgreSQL running on my laptop already.

edsu reviewed (approved):
I think we need to add psycopg2 to the dependencies for the alembic migration to work?

Also maybe we could remove docker-compose.yml if the Mongo config is no longer needed. I was super confused at first when the Postgres connection failed when I ran uv run alembic upgrade head and got a connection error. I thought maybe Mongo was running, but in reality the Dockerized PostgreSQL from the compose.yml was running it just couldn't bind to the port as I had PostgreSQL running on my laptop already.

I'm marking this as approved though! 🚀
